apps/glusterfs/operations_block_volume.go

```diff
@@ -376,6 +376,264 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+func loadBlockVolumeExpandOperation(
+	db wdb.DB, p *PendingOperationEntry) (*BlockVolumeExpandOperation, error) {
+
+	bvs, err := blockVolumesFromOp(db, p)
+	if err != nil {
+		return nil, err
+	}
+	if len(bvs) != 1 {
+		return nil, fmt.Errorf(
+			"Incorrect number of block volumes (%v) for expand operation: %v",
+			len(bvs), p.Id)
+	}
+
+	return &BlockVolumeExpandOperation{
+		OperationManager: OperationManager{
+			db: db,
+			op: p,
+		},
+		bvol: bvs[0],
+	}, nil
+}
+
+func (bve *BlockVolumeExpandOperation) Label() string {
+	return "Expand Block Volume"
+}
+
+func (bve *BlockVolumeExpandOperation) ResourceUrl() string {
+	return fmt.Sprintf("/blockvolumes/%v", bve.bvol.Info.Id)
+}
+
+func (bve *BlockVolumeExpandOperation) Build() error {
+	return bve.db.Update(func(tx *bolt.Tx) error {
+		txdb := wdb.WrapTx(tx)
+		v, err := NewBlockVolumeEntryFromId(tx, bve.bvol.Info.Id)
+		if err != nil {
+			return err
+		}
+		bve.bvol = v
+		if bve.bvol.Pending.Id != "" {
+			logger.LogError("Pending block volume %v can not be Expanded",
+				bve.bvol.Info.Id)
+			return ErrConflict
+		}
+		err = bve.bvol.expandBlockVolumeComponents(txdb, bve.newSize)
+		if err != nil {
+			return err
+		}
+
+		bve.op.RecordExpandBlockVolume(bve.bvol, bve.newSize)
+		if e := bve.op.Save(tx); e != nil {
+			return e
+		}
+		if e := bve.bvol.Save(tx); e != nil {
+			return e
+		}
+		return nil
+	})
+}
+
+func (bve *BlockVolumeExpandOperation) Exec(executor executors.Executor) error {
+	var (
+		err     error
+		bv      *BlockVolumeEntry
+		hvname  string
+		bvHosts nodeHosts
+	)
+	err = bve.db.View(func(tx *bolt.Tx) error {
+		txdb := wdb.WrapTx(tx)
+		bv, err = NewBlockVolumeEntryFromId(tx, bve.bvol.Info.Id)
+		if err != nil {
+			return err
+		}
+		hvname, err = bv.blockHostingVolumeName(txdb)
+		if err != nil {
+			return err
+		}
+		bvHosts, err = bv.hosts(txdb)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		logger.LogError(
+			"failed to get state needed to expand block volume: %v", err)
+		return err
+	}
+	// nothing past this point needs a db reference
+	logger.Info("executing expand of block volume %v in op:%v",
+		bve.bvol.Info.Id, bve.op.Id)
+	return newTryOnHosts(bvHosts).once().run(func(h string) error {
+		return bv.expandBlockVolumeExec(executor, hvname, bve.newSize, h)
+	})
+}
+
+func (bve *BlockVolumeExpandOperation) Rollback(executor executors.Executor) error {
+	logger.Info("Starting Rollback for %v op:%v", bve.Label(), bve.op.Id)
+	return rollbackViaClean(bve, executor)
+}
+
+func (bve *BlockVolumeExpandOperation) Finalize() error {
+	return bve.db.Update(func(tx *bolt.Tx) error {
+
+		// block volume properties are mutated by the results coming
+		// back during exec. These properties need to be saved back
+		// to the db.
+		// This is only noteworthy because it is different from regular
+		// volumes which determines everything up front. Here certain
+		// values are determined by gluster-block commands.
```

Comment:
We're probably copying the code for file volume expand a bit too closely here. This isn't the best pattern, at least as far as mutating the entry objects outside of a transaction and then expecting consistent results afterward.
A better pattern would be to capture the results needed from the system in the expand exec function in an object(s) or attributes specifically created to track what the state of the system is rather than piggyback on db entries.
This also makes it cleaner wrt removing the *<x>Entry as an argument to initialize the operation.

Reply:
John,

I have copied this piece of comments intentionally from the same file, it is part of function:
```
func (bvc *BlockVolumeCreateOperation) Finalize() error {
    ...
}
```

I was actually unclear with what we wanted to do to fix completely.

Hence, I would like to let this PR carry as it is, if its not a serious problem, and I will definitely address it in both the places in a separate PR.